### ISSUE TITLE
"upgrade --package" must remove old dependencies

### ISSUE THREAD

**Issue by gramakri** [NodeBB/NodeBB#8547]
"upgrade --package" option is used to update an existing installation's package.json to a newer NodeBB version and preserving the plugin module names at the same time. This command currently adds extra deps and preserves plugin module names but does not remove old/obsolete deps.

Related forum post: https://community.nodebb.org/topic/14919/maintaining-package-json-across-updates

**Comment by pitaj** [comment]
I'm fairly certain it never simultaneously preserved plugins and removed old deps. We will fix this though as discussed on the forum.

**Comment by barisusakli** [comment]
What happens if someone does `npm i some-module --save` that is not in install/package.json. Is it gonna get removed in the next upgrade. If this is fixed then you won't be able to use any custom modules if they are not in install/package.json?

**Comment by julianlam** [comment]
The existing behaviour would not change, the only new effect would be any packages that do not exist in install/package.json AND do not start with `nodebb-` will be removed. This should preserve plugins and themes. There should be no other reason someone would install separate packages to NodeBB...

**Comment by barisusakli** [comment]
Adding newrelic comes to mind, which isn't in install/package.json 🤔

**Comment by pitaj** [comment]
@baris what do you mean by adding newrelic? Wouldn't you need a plugin to do that?

**Comment by julianlam** [comment]
Specifically with newrelic, no you'd add the `require` at the top of `app.js` after installing the dependency... I suppose that's the only edge case, and isn't something we'd want to build around.

**Comment by pitaj** [comment]
If you have to modify source code, you may as well modify install/package.json as well.

**Comment by gramakri** [comment]
I also want to point out the extra deps in package.json is only an issue because NodeBB has some code that checks if all the deps in package.json exist before starting up. This appears to be true by default as well. Is this check actually required (atleast by default)?

**Comment by pitaj** [comment]
I don't see why extra deps in package.json is an issue, even with the dependency checker. They should be installed by npm just like all of the other packages in there. You should never have a case where `node_modules` is out of sync with package.json, as every npm command is now implicitly `--save`.

We could change it to check install/package.json instead, but then you have issues where people can't upgrade or downgrade specific plugins or packages to fix bugs.

The dependency check is there because many dependencies are lazy loaded. A missing dependency could cause an error while NodeBB is running. This can still happen now if somebody forgets to run `./nodebb upgrade` though.

**Comment by gramakri** [comment]
@pitaj You are right that calling npm install will fix the issue of extra deps. Thing is we don't call npm install in Cloudron package and depending on what you guys decide with this issue, I will adjust the package code accordingly.

In Cloudron, we package all apps as read-only containers, which is great from a security perspective and also for code immutability. This then means that we do npm install at on the install/package.json at Docker image build time. When the container runs, at run time, we simple copy over the node modules in the docker image to a writable directory (which now doesn't contain the old deps of previous nodebb versions). Actually, we have to do all this copying only because we have to install plugins in node_modules (we do this using "nodebb plugins" output). [1] Like I mentioned in the start, I can modify the start script to do npm install on every startup but the only reason to do this is to install obsolete modules and just keep the dep checker happy :-)

[1] Ideally, if nodebb could load modules from some other special directory like NODEBB_PLUGINS_DIR or something, we can avoid this whole copying and just load up entire NodeBB code with modules straight from docker image. This is what we do for most other apps. But this is a separate issue and if you like this idea, I am happy to open a separate report. I don't even know if it's possible to load up node modules from arbitrary folder like this.

**Comment by pitaj** [comment]
Okay so I've created a fix for this issue.

I'm still a little confused as to what your setup is. Do you support adding plugins at runtime? If so, you must allow for the `node_modules` directory, the `build` directory, and `package.json` to be modified at runtime. And you still have to install the active plugins at startup, right?

If my assumptions are correct, you could modify your startup to copy `install/package.json` to `package.json` and then just call `npm install [active plugins]` (`--save` is implicit now) and you'd be good. What am I missing?

> Ideally, if nodebb could load modules from some other special directory like NODEBB_PLUGINS_DIR or something, we can avoid this whole copying and just load up entire NodeBB code with modules straight from docker image. This is what we do for most other apps. But this is a separate issue and if you like this idea, I am happy to open a separate report. I don't even know if it's possible to load up node modules from arbitrary folder like this.

Now it's kinda sounding like you don't allow for installing plugins at runtime. Why can't you just create a `node_modules` symlink to wherever these packages are stored?

Or are you saying that you want node to load packages from two different directories? There's a way to do that, if you disable the dep checker (just put `"dep-check": false` in config.json). Say you have nodebb installed at `/home/nodebb/nodebb`. You can symlink your read-only package directory to `/home/nodebb/node_modules` and node should load from there when it starts up.

**Comment by gramakri** [comment]
Thanks! I will update Cloudron's NodeBB package accordingly.
